doc/tag.1
.TH TAG 1 "17 Jan 2018" "0.1"
.SH NAME
tag \- a tool for using tags in files
.SH SYNOPSIS
.B tag
.RI [ help | -h | --help ]
.PP
.B tag
.RI [ version | -v | --version ]
.PP
.B tag
.IR add | filter | remove
TAGS [PATH]...
.PP
.B tag
.I clear
[DIRECTORY|.tags]...
.PP
.B tag
.I find
TAGS [DIRECTORY]...
.PP
.B tag
.I list
[PATH]...
.PP
.B tag
.IR copy | move
SRC DEST
.SH DESCRIPTION
Tag allows a better file organization, with the use of short descriptions (tags)
to group related files.
.SH OPTIONS
Each command interact with files differently, but their options follows the
same idea:
.PP
.TP
.I TAGS
A list of comma (,) separated tags, with no space after the comma, but
whitespace inside the tag is allowed (you will have to quote the argument). If
it is only one tag, there shouldn't exist any comma.
.PP
.TP
.I PATH
The path to a file, absolute or relative to the current working directory. It
can be provided from
.BR stdin .
.PP
.TP
.I DIRECTORY
The same as
.IR PATH ,
but must be a valid directory.
.PP
.TP
.I .tags
The same as
.IR PATH ,
but must be a valid
.I .tags
file.
.PP
.TP
.I SRC
The same as
.IR PATH ,
but must be specified as an argument, and not from
.BR stdin .
.PP
.TP
.I DEST
The same as
.IR SRC .
.SH COMMANDS
.BR help ", " -h ", " --help
.RS
Outputs the
.BR SYNOPSIS /USAGE.
.RE
.PP
.BR version ", " -v ", " --version
.RS
Outputs the version.
.RE
.PP
.B add
.IR TAGS " [" PATH ]...
.RS
Adds
.I TAGS
into the
.I .tags
relative to each
.IR PATH .
Duplicate tags are removed.
.RE
.PP
.B filter
.IR TAGS " [" PATH ]...
.RS
Echos only
.I PATH
which contain all
.IR TAGS .
.RE
.PP
.B remove
.IR TAGS " [" PATH ]...
.RS
Removes all
.I TAGS
from all
.IR PATH .
.I PATH
does not need to exist, but it relative
.I .tags
does.
.RE
.PP
.B clear
.RI [ DIRECTORY | .tags ]...
.RS
Removes file entries in
.I .tags
that do not exist. If a
.I DIRECTORY
is provided, all
.I .tags
inside it (recursively) are cleared.
.RE
.PP
.B find
.IR TAGS " [" DIRECTORY ]...
.RS
Finds all files inside
.I DIRECTORY
(recursively) that have all
.IR .tags .
Paths returned are relative to
.IR DIRECTORY .
.RE
.PP
.B list
.RI [ PATH ]...
.RS
Lists tags associated to
.IR PATH ,
in a comma separated list.
.RE
.PP
.B copy
.I SRC DEST
.RS
Copies
.I SRC
tags to
.IR DEST ,
preserving previous
.I DEST
tags. If
.I DEST
does not exist,
.I SRC
is copied into it. See
.I -T
in
.BR cp (1).
.RE
.PP
.B move
.I SRC DEST
.RS
Moves
.I SRC
tags to
.IR DEST ,
preserving previous
.I DEST
tags and removing all
.I SRC
tags. If
.I DEST
does not exist,
.I SRC
is moved into it. See
.I -T
in
.BR mv (1).
.RE
.SH ERROR CODES
.IP 1
Bad command call
.IP 2
Unexpected argument
.IP 3
File does not exist or is inaccessible
.IP 4
Not a directory or
.I 3
.IP 5
File has no tags
.IP 6
File is not a .tags file
.SH SEE ALSO
.BR awk (1),
.BR basename (1),
.BR bash (1)
.IR command ,
.BR cp (1),
.BR cut (1),
.BR dirname (1),
.BR find (1),
.BR grep (1),
.BR mv (1),
.BR printf (1),
.BR sed (1)
.SH BUGS
.IR PATH " and " TAGS
can not contain "\\n".
.PP
If you add a tag in the current directory, the
.I .tags
file is created there,
instead of in the parent directory.
.SH NOTES
Developed with GNU bash, version
.I 4.3.48(1)-release
.SH AUTHOR
Aryel Mota Góis
.RB ( aryelgois
at
.IR GitHub )
.SH COPYRIGHT
Copyright (c) 2018 Aryel Mota Góis. MIT License.


src/tag.sh
#!/usr/bin/env bash
#
# tag - a tool for using tags in files
#
# @author Aryel Mota Góis
# @license MIT
# @link https://github.com/aryelgois/tag

VERSION='0.1'

#
# Utils
#

# Split a string into an array
#   $1 string to be splitted (byref)
#   $2 char delimiter (default to x1F)
function array_parse {
    local -n ARR=$1
    local IFS=${2:-$'\x1f'}
    ARR=($ARR)
}

# Remove all occurrences of a value in an array
#   $1 array (byref)
#   $2 value to be removed
function array_remove {
    local -n ARR=$1
    local DELETE=$2
    local TMP=()
    for i in "${ARR[@]}"; do [[ $i != $DELETE ]] && TMP+=($i); done
    ARR=("${TMP[@]}")
}

# Join an array items with a single character
#   $1 array to be joined (byref)
#   $2 char delimiter (default to x1F)
function array_stringify {
    local -n ARR=$1
    local IFS=${2:-$'\x1f'}
    ARR=("${ARR[*]}")
}

# Remove repeated items in array
#   $1 array (byref)
#   $2 char delimiter used during array rebuild (default to x1F)
#
# NOTE:
#   - It removes null values
#   - Result is unordered
function array_unique {
    local -n TMP1=$1
    local -A TMP2
    local SEP=${2:-$'\x1f'}
    SEP="${SEP::1}"
    for i in "${TMP1[@]}"; do [[ -n $i ]] && TMP2["$i"]=1; done
    TMP1="$(printf "%s$SEP" "${!TMP2[@]}")"
    array_parse TMP1 $SEP
}

# Get FILE array from arguments or stdin
#   $@ files
#
# If any argument is provided, FILES will contain them. Otherwise, FILES will
# contain the stdin, if it's not a TTY
function get_files {
    FILES=()
    if [[ $# -gt 0 ]]; then
        FILES=("$@")
    elif [[ ! -t 0 ]]; then
        FILES="$(</dev/stdin)"
        array_parse FILES $'\n'
    fi
}

# Echo to stderr
#   $@ output message
function errcho {
    >&2 echo "$@"
}

# Test if file exists
#   $1 path to file
#
# Error code:
#   3 file does not exist or is inaccessible
function file_exists {
    if [[ ! -e $1 ]]; then
        errcho "E: File '$1' does not exist or is inaccessible"
        return 3
    fi
}

# Test if directory exists
#   $1 path to directory
#
# Error code:
#   4 not a directory or 3 in file_exists()
function dir_exists {
    if [[ ! -d $1 ]]; then
        errcho "E: '$1' is not a directory, does not exist or is inaccessible"
        return 4
    fi
}

# Test if file has tags and stdout them
#   $1 path to file
#
# Error code:
#   5 file has no tags
function has_tag {
    local TAGS="$(list "$1")"
    if [[ -z $TAGS ]]; then
        errcho "E: '$1' has no tags"
        return 5
    fi
    echo $TAGS
}

# Escape string to embed in grep
#   $* string to be escaped
function ere_quote {
    sed 's/[][\/\.|$(){}?+*^]/\\&/g' <<< "$*"
}

# Get tags associated to a file in a .tags
#   $1 path to .tags
#   $2 file name to
function read_tags {
    local TMP=
    if [[ -e $1 ]]; then
        TMP=$(grep "^$(ere_quote "$2")/" "$1" | tail -n 1)
        TMP="${TMP:((${#2} + 1))}"
    fi
    echo $TMP
}

#
# Commands
#

# Output short help about the script
#   $1 status code to return
function help {
    echo "\
USAGE:
    tag [help|-h|--help]

    tag [version|-v|--version]

    tag add|filter|remove TAGS [PATH]...

    tag clear [DIRECTORY|.tags]...

    tag find TAGS [DIRECTORY]...

    tag list [PATH]...

    tag copy|move SRC DEST

PATH and DIRECTORY can be passed with stdin

For details, see tag(1)"
    exit $1
}

# Add a tag to a file
#   $1 path to file
#   $2 tags to add
#
# NOTE:
#   - Tags are listed in '.tags' at the file's directory
function add {
    file_exists "$1" || return

    local BASENAME=$(basename "$1")
    local TAG_FILE="$(dirname "$1")/.tags"
    local TAGS="$(read_tags "$TAG_FILE" "$BASENAME"),$2"

    array_parse TAGS ,
    array_unique TAGS ,
    array_stringify TAGS ,

    [[ -e $TAG_FILE ]] && sed -i "/^$(ere_quote "$BASENAME")\// d" "$TAG_FILE"
    echo "$BASENAME/$TAGS" >> "$TAG_FILE"
}

# Filter file with a tag
#   $1 path to file
#   $2 tags to match
#
# stdout $1 if it has all tags
function filter {
    file_exists "$1" || return

    local BASENAME=$(basename "$1")
    local TAG_FILE="$(dirname "$1")/.tags"
    local TAGS="$(ere_quote "$2")"
    local LIST="/^$(ere_quote "$BASENAME")\//"
    local FOUND=

    array_parse TAGS ,
    for i in "${TAGS[@]}"; do
        LIST="$LIST && /(^[^\/]*\/|,)$i(,|$)/"
    done

    FOUND=$(awk "$LIST" "$TAG_FILE")

    [[ -n $FOUND ]] && echo "$1"
}

# Find files with a tag inside a directory
#   $1 path to directory
#   $2 tags to search
#
# stdout list of files found with all tags
function find {
    dir_exists "$1" || return

    local TAGS="$(ere_quote "$2")"
    local LIST=

    array_parse TAGS ,
    for i in "${TAGS[@]}"; do
        LIST="$LIST && /(^[^\/]*\/|,)$i(,|$)/"
    done
    LIST="${LIST:4}"

    command find "$1" -type f -name .tags | while read TAG_FILE; do
        DIRNAME=$(dirname "$TAG_FILE")
        awk "$LIST" "$TAG_FILE" | cut -d / -f 1 | while read FILE; do
            echo "$DIRNAME/$FILE"
        done
    done
}

# Remove tags associated to a file
#   $1 path to file
#   $2 tags to remove
#
# NOTE:
#   - The file does not need to exist, but the .tags file does
function remove {
    local TAG_FILE="$(dirname "$1")/.tags"
    file_exists "$TAG_FILE" || return

    local BASENAME=$(basename "$1")
    local TAGS=$(read_tags "$TAG_FILE" "$BASENAME")
    local REMOVE="$2"

    array_parse TAGS ,
    array_parse REMOVE ,
    for i in "${REMOVE[@]}"; do
        array_remove TAGS "$i"
    done
    array_stringify TAGS ,

    sed -i "/^$(ere_quote "$BASENAME")\// d" "$TAG_FILE"

    if [[ -n $TAGS ]]; then
        echo "$BASENAME/$TAGS" >> $TAG_FILE
    elif [[ ! -s $TAG_FILE ]]; then
        rm "$TAG_FILE"
    fi
}

# Clear .tags files listing nonexistent files
#   $1 path to directory or .tags file
#
# Error code:
#   6 file is not a .tags file
#
# NOTE:
#   - If a directory is passed, all .tags inside it are cleared
function clear {
    file_exists "$1" || return

    local BASENAME=$(basename "$1")

    if [[ -d $1 ]]; then
        command find "$1" -type f -name .tags | while read TAG_FILE; do
            clear "$TAG_FILE"
        done
        return
    elif [[ $BASENAME != '.tags' ]]; then
        errcho "E: File '$1' is not a .tags file"
        return 6
    fi

    local DIRNAME=$(dirname "$1")

    cat "$1" | cut -d / -f 1 | while read FILE; do
        [[ -e "$DIRNAME/$FILE" ]] || sed -i "/^$(ere_quote "$FILE")\// d" "$1"
    done

    [[ ! -s $1 ]] && rm "$1"
}

# List file tags
#   $1 path to file
function list {
    file_exists "$1" || return

    local BASENAME=$(basename "$1")
    local TAG_FILE="$(dirname "$1")/.tags"
    local TAGS=$(read_tags "$TAG_FILE" "$BASENAME")

    array_parse TAGS ,
    array_unique TAGS ,
    array_stringify TAGS ,

    echo "$TAGS"
}

# Copy tags from one file to another
#   $1 path to source file
#   $2 path to destiny file
#
# Error code:
#   5 source has no tags
#
# NOTE:
#   - Also copy the file if destiny does not exist
function copy {
    file_exists "$1" || return

    local BASENAME=$(basename "$2")
    local TAG_FILE="$(dirname "$2")/.tags"
    local TAGS="$(has_tag "$1")"

    [[ -n $TAGS ]] || return 5

    [[ -e $2 ]] || cp -T "$1" "$2"

    add "$2" "$TAGS"
}

# Move tags from one file to another
#   $1 path to source file
#   $2 path to destiny file
#
# Error code:
#   5 source has no tags
#
# NOTE:
#   - Also move the file if destiny does not exist
function move {
    file_exists "$1" || return

    local BASENAME=$(basename "$2")
    local TAG_FILE="$(dirname "$2")/.tags"
    local TAGS="$(has_tag "$1")"

    [[ -n $TAGS ]] || return 5

    remove "$1" "$TAGS"

    [[ -e $2 ]] || mv -T "$1" "$2"

    add "$2" "$TAGS"
}

#
# Main
#

if [[ $# -gt 0 ]]; then
    KEY="$1"
    NEXT="$2"
    shift
    case $KEY in
    help-h|--help)
        help
        ;;
    version|-v|--version)
        echo "\
tag - a tool for using tags in files
version: $VERSION"
        exit
        ;;
    add|filter|find|remove)
        [[ -n $NEXT ]] && shift || help 1

        get_files "$@"
        if [[ ${#FILES[@]} -eq 0 ]]; then
            [[ $KEY == 'find' ]] && FILES=('.') || help 1
        fi

        for FILE in "${FILES[@]}"; do
            $KEY "$FILE" "$NEXT"
        done
        exit
        ;;
    clear|list)
        get_files "$@"
        if [[ ${#FILES[@]} -eq 0 ]]; then
            [[ $KEY == 'clear' ]] && FILES=('.') || help 1
        fi

        for FILE in "${FILES[@]}"; do
            $KEY "$FILE"
        done
        exit
        ;;
    copy|move)
        if [[ $# -eq 2 ]]; then
            $KEY "$1" "$2"
        else
            help 1
        fi
        exit
        ;;
    *)
        errcho "E: unexpected argument '$KEY'"
        exit 2
        ;;
    esac
fi

help 1
